# Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

data_path = "data/supermarket_sales.csv"
supermarket_sales_dataset = pd.read_csv(data_path)
supermarket_sales_dataset.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
supermarket_sales_dataset.groupby("Branch")["gross income"].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
supermarket_sales_dataset.groupby(["City", "Branch"])["gross income"].sum()

City       Branch
Mandalay   B         5057.0320
Naypyitaw  C         5265.1765
Yangon     A         5057.1605
Name: gross income, dtype: float64

In [5]:
# Calculate total gross income by branch
total_gross_income_by_branch = supermarket_sales_dataset.groupby(['Branch', 'City'])['gross income'].sum()

# Find the branch with the highest gross income
highest_gross_income_branch = total_gross_income_by_branch.idxmax()
highest_gross_income_branch_total = total_gross_income_by_branch.max()

highest_gross_income_branch, highest_gross_income_branch_total


(('C', 'Naypyitaw'), np.float64(5265.1765))

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [6]:
# supermarket_sales_dataset.groupby(["Product line", "City"])["Branch"].count()

invoices_by_category = supermarket_sales_dataset.pivot_table(
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count',
    fill_value=0
)

invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [7]:
branch_max_electronic = invoices_by_category['Electronic accessories'].idxmax()
max_invoices_electronic = invoices_by_category['Electronic accessories'].max()

branch_max_electronic, max_invoices_electronic

('A', np.int64(60))

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


<h1 style='font-family: Garamond ;'><u>Гіпотеза 1</u>: Який середній чек (<b>Average Order Value</b>, <b>AOV</b>) та кількість товарів у чеку (<b>Items per Purchase</b>) в розрізі категорій товарів (Product Line) і типів клієнтів (Customer Type)?</h1>
<hr></hr>

In [8]:
product_metrics = pd.pivot_table(
    supermarket_sales_dataset,
    values=['Total', 'Quantity'],
    index='Product line',
    columns='Customer type',
    aggfunc='mean'
).round(2)

product_metrics

Quantity          Total        
Customer type            Member Normal  Member  Normal
Product line                                          
Electronic accessories     5.50   5.89  314.08  324.34
Fashion accessories        5.10   5.03  306.09  304.15
Food and beverages         5.38   5.58  333.59  309.84
Health and beauty          5.86   5.39  353.85  295.73
Home and lifestyle         5.90   5.47  337.08  336.15
Sports and travel          5.67   5.41  324.53  340.36

<p style='font-family: Garamond ; font-size: 18px;'>Маємо наступні категоріальні значення за розподілами у зведеній таблиці:</p>
<ul style='font-family: Garamond ; font-size: 18px;'>
    <li><code>Total</code> = середній чек для кожної категорії відповідно до типу клієнта.</li>
    <li><code>Quantity</code> = скільки одиниць товару купують в середньому.</li>
</ul>
<ul style='font-family: Garamond ; font-size: 18px;'>Розглянемо категорію від <code>Customer Type</code> Member у контексті <code>Quantity</code> та <code>Total</code>:
    <li>Максимальний AOV - категорія Health and Beauty.</li>
    <li>Максимальний IPP - категорія Home and Lifestyle, також дуже близько Health and Beauty.</li>
</ul>
<ul style='font-family: Garamond ; font-size: 18px;'>Розглянемо категорію від <code>Customer Type</code> Normal у контексті <code>Quantity</code> та <code>Total</code>:
    <li>Максимальний AOV - категорія Sports and Travel.</li>
    <li>Максимальний IPP - категорія Electronic Accessories.</li>
</ul>
<p style='font-family: Garamond ; font-size: 18px;'>Ці категорії мають потенціал для створення пакетних пропозицій (<b>bundle</b>). Вартість доставки може бути розрахована гнучко згідно з типовим об'ємом замовлення, а також середнім чеком. Наприклад, від $340 безкоштовна доставка.</p>

<h1 style='font-family: Garamond ;'><u>Гіпотеза 2</u>: Який рівень апсейлу (<b>Upsell Rate</b>) — співвідношення суми податку (<code>Tax 5%</code>) до кількості одиниць товару (<code>Quantity</code>) — найвищий для яких категорій?</h1>
<hr></hr>

In [9]:
supermarket_sales_dataset['Upsell Rate'] = supermarket_sales_dataset['Tax 5%'] / supermarket_sales_dataset['Quantity']
upsell_by_category = supermarket_sales_dataset.groupby('Product line')['Upsell Rate'].mean().sort_values(ascending=False).reset_index()

upsell_by_category.head(3)

Product line  Upsell Rate
0  Fashion accessories     2.857683
1    Sports and travel     2.849663
2   Food and beverages     2.800443

<p style='font-family: Garamond ; font-size: 18px;'>Найвищий <b>податок на одиницю</b> — своєрідний індикатор того, що ці товари або дорожчі, або з високою маржою, і їх вигідно було б додати у bundle до основних позицій. Крім того, ми прослідковуємо факт того, що топ-3 категорії досить комбінативні для інших категорій (менш прибуткових), аби підвищити середню вартість чеку.</p>

<ul style='font-family: Garamond ; font-size: 18px;'><b>Рекомендації</b>:
    <li>Вартість апсейлу може бути використана для створення стратегій <b>upsell/cross-sell</b>.</li>
    <li>Найвищий показник у <code>Fashion accessories</code> — ідеальний кандидат для аксесуарів до дорогих основних товарів.</li>
</ul>


<h1 style='font-family: Garamond ;'><u>Гіпотеза 3</u>: Які категорії товарів мають найвищу конверсію в прибуток: Gross Margin % × Volume (кількість)?</h1>
<hr></hr>

In [11]:
supermarket_sales_dataset['Margin Volume'] = supermarket_sales_dataset['gross margin percentage'] * supermarket_sales_dataset['Quantity']
margin_volume_by_category = supermarket_sales_dataset.groupby('Product line')['Margin Volume'].sum().sort_values(ascending=False).reset_index()
margin_volume_by_category

Product line  Margin Volume
0  Electronic accessories    4623.809524
1      Food and beverages    4533.333333
2       Sports and travel    4380.952381
3      Home and lifestyle    4338.095238
4     Fashion accessories    4295.238095
5       Health and beauty    4066.666667

<p style='font-family: Garamond ; font-size: 18px;'>Це поєднання маржинальності та об’єму продажів — ключовий показник ефективності категорії. Топові категорії варто просувати через масові акції (сезонні (якщо мова йде про <code>Electronic accessories</code>, то взимку можна просувати обігрівачі, навесні - розумні годинники та фітнес-браслети(початок активного періоду спорту простонеба - біг, фітнес, йога на узбережжі та впарку), улітку - кондиціонери та вентилятори, восени - електроковдри, електронні каміни, кавоварки тощо)) — вони тягнуть загальний прибуток. У плані розробки акцій додатково можна дослідити розподіл прибутку категорії товарів відповідно до розподілу за місяцями.</p> 